crates/corro-agent/src/agent/util.rs

```diff
@@ -0,0 +1,1479 @@
+                    .layer(HandleErrorLayer::new(|_error: BoxError| async {
+                        Ok::<_, Infallible>((
+                            StatusCode::SERVICE_UNAVAILABLE,
+                            "max concurrency limit reached".to_string(),
+                        ))
+                    }))
+                    .layer(LoadShedLayer::new())
+                    .layer(ConcurrencyLimitLayer::new(128)),
+            ),
+        )
+        .route(
+            "/v1/migrations",
+            post(api_v1_db_schema).route_layer(
+                tower::ServiceBuilder::new()
+                    .layer(HandleErrorLayer::new(|_error: BoxError| async {
+                        Ok::<_, Infallible>((
+                            StatusCode::SERVICE_UNAVAILABLE,
+                            "max concurrency limit reached".to_string(),
+                        ))
+                    }))
+                    .layer(LoadShedLayer::new())
+                    .layer(ConcurrencyLimitLayer::new(4)),
+            ),
+        )
+        .layer(axum::middleware::from_fn(require_authz))
+        .layer(
+            tower::ServiceBuilder::new()
+                .layer(Extension(Arc::new(AtomicI64::new(0))))
+                .layer(Extension(agent.clone()))
+                .layer(Extension(subs_bcast_cache))
+                .layer(Extension(subs_manager.clone()))
+                .layer(Extension(tripwire.clone())),
+        )
+        .layer(DefaultBodyLimit::disable())
+        .layer(TraceLayer::new_for_http());
+
+    let api_addr = api_listener.local_addr()?;
+    info!("Starting public API server on tcp/{api_addr}");
+    spawn_counted(
+        axum::Server::builder(AddrIncoming::from_listener(api_listener)?)
+            .executor(CountedExecutor)
+            .serve(
+                api.clone()
+                    .into_make_service_with_connect_info::<SocketAddr>(),
+            )
+            .with_graceful_shutdown(
+                tripwire
+                    .clone()
+                    .inspect(move |_| info!("corrosion api http tripped {api_addr}")),
+            )
+            .inspect(|_| info!("corrosion api is done")),
+    );
+
+    Ok(())
+}
+
+async fn require_authz<B>(
+    Extension(agent): Extension<Agent>,
+    maybe_authz_header: Option<TypedHeader<Authorization<Bearer>>>,
+    request: axum::http::Request<B>,
+    next: axum::middleware::Next<B>,
+) -> Result<axum::response::Response, axum::http::StatusCode> {
+    let passed = if let Some(ref authz) = agent.config().api.authorization {
+        match authz {
+            AuthzConfig::BearerToken(token) => maybe_authz_header
+                .map(|h| h.token() == token)
+                .unwrap_or(false),
+        }
+    } else {
+        true
+    };
+
+    if !passed {
+        return Err(axum::http::StatusCode::UNAUTHORIZED);
+    }
+
+    Ok(next.run(request).await)
+}
+
+// DOCME: provide some context for this function
```

Comment:
This is a fun one.

When we insert `crsql_changes` we get via broadcasts and sync process, we don't know which changes from the past might've been invalidated (cleared). Sometimes whole versions are cleared and we don't know until we make these expensive queries (they're indexed at least).

When we don't know about cleared versions, we accumulate much more data in the `__corro_bookkeeping` table and, more importantly, the structure in-memory for "current" versions (not cleared) is a lot bigger. Cleared versions are stored both in the DB and in-memory as ranges like "versions `1..=5` for actor-id `x` don't appear in the DB anymore" and therefore only use minimal space.

If we don't find these cleared versions periodically, memory keeps growing as more changes are processed (especially bad for many small changes being processed). Finding them in the DB lets us rewrite the bookkeeping DB in a more compact format and the same goes for the in-memory DB.